### ISSUE TITLE
flutter run should support wireless debugging of iOS devices

### ISSUE THREAD

**Issue by phumberdroz** [flutter/flutter#15072]
Latest update: https://github.com/flutter/flutter/issues/15072#issuecomment-898202823

----

## Steps to Reproduce

connect iPhone via usb
Open xcode -> window -> devices & Simulators -> select correct device and enable connect via network
Network Symbol is shown in the list to the left
disconnect iPhone

Android Studio shows devices but if you try to run it it is in an endless waiting loop without visual indication 

If you then connect the devices via USB it continues on

## Logs
https://gist.github.com/phumberdroz/9700d43d2bc9f5e510bebc6dab9b509c
## Flutter Doctor

```
[✓] Flutter (Channel beta, v0.1.5, on Mac OS X 10.13.3 17D102, locale en-US)
    • Flutter version 0.1.5 at /Users/phumberdroz/flutter
    • Framework revision 3ea4d06340 (8 days ago), 2018-02-22 11:12:39 -0800
    • Engine revision ead227f118
    • Dart version 2.0.0-dev.28.0.flutter-0b4f01f759

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/phumberdroz/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)

[✓] iOS toolchain - develop for iOS devices (Xcode 9.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.2, Build version 9C40b
    • ios-deploy 1.9.2
    • CocoaPods version 1.4.0

[✓] Android Studio (version 3.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)

[✓] IntelliJ IDEA Ultimate Edition (version 2017.3.4)
    • Flutter plugin version 22.2.2
    • Dart plugin version 173.4548.30

[✓] VS Code (version 1.20.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Dart Code extension version 2.9.2

[✓] Connected devices (2 available)
    • Pierre’s iPhone • ec754a5c0607923df2a99ccd051938e0045dee6b • ios • iOS 11.2.6
    • iPhone 7        • 5001A557-2FB6-49EB-8C5C-ABC8396F19E9     • ios • iOS 11.2 (simulator)

• No issues found!
> For more information about diagnosing and reporting Flutter bugs, please see [https://flutter.io/bug-reports/](https://flutter.io/bug-reports/).
```

**Comment by Hixie** [comment]
cc @devoncarew IDE bug

**Comment by devoncarew** [comment]
This issue was moved to flutter/flutter-intellij#2308

**Comment by devoncarew** [comment]
@phumberdroz, to re-phrase your report:

- connect iPhone via usb
- use an xcode feature to `enable connect via network`
- disconnect iPhone

We then locate the device correctly, but when you try and run, `flutter run` doesn't work with it. It sounds like flutter_tools doesn't work with the network style connection. @cbracken @xster

**Comment by phumberdroz** [comment]
100%

**Comment by cbracken** [comment]
I suspect we'll need to apply patches to our integration with the `libimobiledevice` toolchain. IIRC, we intentionally disabled wifi debugging due to issues a couple of years ago. I'll see if I can dig up the issue(s) and x-ref them here then we can look at a fix. Tooling has improved significantly since then, so definitely worth another look.

**Comment by cbracken** [comment]
#8309 is where I disabled this.

**Comment by tonikami** [comment]
@cbracken Any update on this?

**Comment by RandalSchwartz** [comment]
@crazybit_gitlab on the Flutter Gitter channel reports success with the following:

Correction: I got adb over wifi to work without issue. Flutter works also. Awesome, no need for adapters and dongles after all!

I followed the following instructions: https://developer.android.com/studio/command-line/adb#wireless

Need to follow all steps and it was important to wait a minute or so after disconnecting the USB step, because otherwise adb would show the device as offline.

**Comment by anderscheow** [comment]
We need this :(

**Comment by zoechi** [comment]
The thumbs-ups should be on the initial comment to raise priority.

**Comment by Hixie** [comment]
@anderscheow Can you talk more about why you need this? (It would help us with prioritization.)

**Comment by Hixie** [comment]
cc @cyanglaz

**Comment by anderscheow** [comment]
@Hixie Using traditional way (wired connection) to do debugging is kinda troublesome. At some points, if cable is having issue, I need to get another cable in order to make debugging works.

**Comment by kunit1** [comment]
In my case, one of the test devices has a bad lightning port; the device is otherwise fully functional. This would allow me to use this device for debugging while avoiding the finicky cable.

**Comment by ride4sun** [comment]
I think in general it is a convenience feature which I would expect from a
new platform like flutter.
Hot reload is also not necessary but moving forward developer will expect
it from platforms after flutter because it will become a standard like
wireless is already.

**Comment by Hixie** [comment]
Thanks for the clarifications!

**Comment by cbracken** [comment]
Given the direction things are going (e.g. with headphones), one could also envision a day when Apple ships devices that charge wirelessly and don't include a lightning/USB-C port.

**Comment by zjjt** [comment]
Is it possible to flutter run over wifi on android?

**Comment by matthewlloyd** [comment]
I'll add that I've been testing extensively both with Android, over WiFi, and iOS, tethered with a Lightning cable, and it really does make a difference having wireless support. My iPhone is my daily driver so when I need to test with it, I first have to find it (did I leave it on the kitchen counter?), then find the cable (I unplugged it when I took my laptop to the park, but where is it?), plug it in, then pick up the phone with the cable attached getting in the way of my hand, making it clumsy to test; iTunes automatically opens and I have to close it, then when I'm done debugging, I have to unplug the cable again. It's maybe only a minute of hassle each time, but it adds up, and I find myself testing mostly using WiFi on Android because it's faster and more convenient. It's a bit like hot swap, not strictly necessary but really, really great.

**Comment by takeru** [comment]
External tool libimobiledevice does not supports wifi devices. 🤔 

https://github.com/flutter/flutter/blob/4ff467191b13da6bdb3e0a9420c1d2b487176ed4/packages/flutter_tools/lib/src/ios/mac.dart#L126

https://github.com/libimobiledevice/libimobiledevice/issues/152

**Comment by sionicion** [comment]
I'm purchasing an iOS device and run the actual VSCode on a dedicated server, and code from remote devices. I have to be able to wirelessly debug my apps in this case, and I will be unable to do this until this is fixed. So I would consider this more than just a convenience feature.

**Comment by sionicion** [comment]
I'm doing some experimentation on this because I really need this to work and it looks like wifi support is most likely going to come to the tools Flutter depends on. It appears a one-line change to libimobiledevice is needed to help it identify wifi devices. I also commented out the no-wifi flag in the flutter_tools' devices.dart file. I was then able to see my iPad via wifi and use flutter run. It is able to push the app but I'm currently having this issue.

Cannot debug (mydeviceid) (J321AP, J321AP, uknownos, unkarch)
a.k.a. 'Tyler’s iPad' over WIFI.
[   +3 ms] Application launched on the device. Waiting for observatory port.

I'll see if I can figure out what to do but hopefully someone here can chime in on what's the culprit.

**Comment by sionicion** [comment]
So identifying wifi devices seems pretty easy, it seems it was always able to do it but like it was said up above, it was removed in flutter tools and in the tools Flutter is using.

The tricky part seems to be getting the ports forwarded. I wanted to make sure I didn't have any conflicting duplicate packages installed on my system so I reinstalled Mojave since I wasn't going to stay on Catalina's beta, and I manually built most of the tools Flutter depends on that have pending PRs that help with wifi devices. I wanna say I got a little bit further.

It seems to be trying to forward the ports.

`[  +99 ms] [100%] Installed package build/ios/iphoneos/Runner.app
[ +956 ms] ------ Debug phase ------
[        ] Cannot debug (mydeviceid) (J321AP, J321AP, uknownos, unkarch) a.k.a. 'Tyler’s iPad' over WIFI.
[   +2 ms] Application launched on the device. Waiting for observatory port.
[+7103 ms] Observatory URL on device: http://127.0.0.1:51531/VFV9LsZa8eI=/
[   +1 ms] attempting to forward device port 51531 to host port 1024
[        ] executing: /usr/local/bin/iproxy 1024 51531 00008027-000A41E211E3002E
[+1015 ms] Forwarded port ForwardedPort HOST:1024 to DEVICE:51531
[        ] Forwarded host port 1024 to device port 51531 for Observatory
[   +1 ms] Installing and launching... (completed in 22.7s)
[   +4 ms] Connecting to service protocol: http://127.0.0.1:1024/VFV9LsZa8eI=/
[ +126 ms] Error connecting to the service protocol: HttpException: , uri = http://127.0.0.1:1024/VFV9LsZa8eI=/ws
[   +2 ms] "flutter run" took 32,304ms.
[        ] "flutter run" took 32,304ms.
[   +8 ms] executing: [/Users/me/Apps/myapp/ios/Runner.xcodeproj/] /usr/bin/xcodebuild -project /Users/me/Apps/myapp/ios/Runner.xcodeproj -target Runner -showBuildSettings
[+1040 ms] Exit code 0 from: /usr/bin/xcodebuild -project /Users/me/Apps/myapp/ios/Runner.xcodeproj -target Runner -showBuildSettings
[        ] Build settings for action build and target Runner:`

I'm finishing up playing around with this for tonight. I don't think it should be too difficult to actually get wifi debugging working with iOS, since it's already trying to forward the ports.

**Comment by sarinupreti** [comment]
any updates ?

**Comment by sionicion** [comment]
I gave up because it appeared it wasn't able to forward the ports. It only worked once I plugged my iPad in, once it was disconnected physically, it wasn't able to forward the ports any longer. This all depends on the libimobiledevice's tools. It seems some work is being done on wifi connections but the devs are quiet and don't discuss how their work is going.

**Comment by DanTup** [comment]
@cbracken I've come across a lot of issues recently where people have to disable WiFi on their phone to be able to use Flutter (most recent example https://github.com/flutter/flutter/issues/38971#issuecomment-526727740). It seems that unticking "Connect via network" doesn't help. Is that what https://github.com/flutter/flutter/pull/8309 is intended to help with? Should I close issues like that as dupes of this one?

**Comment by edgarfroes** [comment]
> @cbracken I've come across a lot of issues recently where people have to disable WiFi on their phone to be able to use Flutter (most recent example [#38971 (comment)](https://github.com/flutter/flutter/issues/38971#issuecomment-526727740)). It seems that unticking "Connect via network" doesn't help. Is that what #8309 is intended to help with? Should I close issues like that as dupes of this one?

no sir, this issue is about debugging without the USB cable, over wifi only, so it's a feature. The #38971 issue is a bug when Wifi is enabled on the iOS device when running an app over the USB cable.

**Comment by jmagman** [comment]
> I suspect we'll need this sooner or later.
> 
> Given the direction things are going (e.g. with headphones), one could also envision a day when Apple ships devices that charge wirelessly and don't include a lightning/USB-C port. FWIW I have no actual evidence for this; this is pure speculation on my part.

Already done with the Apple TV 4K.

**Comment by MichaelMarner** [comment]
@Hixie - A major use case for this is when interfacing with external hardware via USB / Lightning. In my case, I am interfacing with external MIDI devices. The app doesn't really function without having the device connected, which means there's no way to have both the hardware connected and debug the app at the same time.

**Comment by edgarfroes** [comment]
> @Hixie - A major use case for this is when interfacing with external hardware via USB / Lightning. In my case, I am interfacing with external MIDI devices. The app doesn't really function without having the device connected, which means there's no way to have both the hardware connected and debug the app at the same time.

Until there's a FIX, you might want to find an USB extender, maybe? Also, test it before buying just so you can make sure iOS can compile properly, they are picky when it comes to the hardware you can use with their products.

**Comment by sooxiaotong** [comment]
well, wait until jailbreak for iOS 12.4 came out.

**Comment by infinito84** [comment]
@cbracken there's a way we can enable the wifi debug? or thus xcode recognizes the device (over wifi), launch  a debug mode of the flutter app from xcode?

**Comment by zxl777** [comment]
Any updates about wifi remote debug with iOS device ?

**Comment by zanderso** [comment]
There's been some work in this area that might affect this issue. Assigning to @jmagman to investigate.

**Comment by jmagman** [comment]
I got as far as @sionicion: can't forward ports.
```
default	16:59:08.696425 -0800	usbmuxd	error     SCEDeviceSocketCallback Client 7451972-iproxy connect to 95-fe80::1c27:e536:56ac:32:0 failed, connection failed with error 61 - Connection refused
```
You can launch the app `ios/Runner.xcworkspace` on a wirelessly paired device in Xcode (or from `flutter` if we decided to swap ideviceinfo and idevicename with some `xcrun instruments -s devices` parsing), but it won't connect to the observatory.

**Comment by SHUBHAMDAVE001** [comment]
How can i connect my Android device with c without using usb cable?

Andriod version 8.1.0

**Comment by jmagman** [comment]
> How can i connect my Android device with c without using usb cable?
> 
> Andriod version 8.1.0
> Here it is not showing option for adb over internet

This issue is tracking iOS issues with wireless debugging.  Please file a new issue if you think there's a gap in Flutter documentation.

**Comment by manojeeva** [comment]
Yes, we need this.
This is going to be very necessary for everyone out there. 
Plugging the USB all the time makes very difficult to test in both android and ios.

**Comment by sc4v3ng3r** [comment]
We need this guys.

**Comment by erf** [comment]
I support this, but my experience is that deploying the app to a device over wifi is very slow, compared to cable.

**Comment by umerwaqas92** [comment]
> Is it possible to flutter run over wifi on android?

Yes it possible, https://stackoverflow.com/a/10236938/4993045

**Comment by Gebes** [comment]
Any Update?
If not, we still need this!

When I code over my MacBook and I build via Flutter (connect through Lightning),
my USB adapter gets hot (due to the power consumption) and it takes tons of battery (from my MacBook) which makes it hard to develop on the go :c

I made it work on my Android, yeah it is slower but you do not rebuilt each time because you can use "hot reload".

**Comment by jmagman** [comment]
Extensive progress has been made to lay the groundwork for this getting switched on.

- Migration off `ideviceinfo` and `idevice_id` to Xcode's device detection CLI that can handle wireless devices:
#49854

- Pipe through longer timeouts for device discovery:
#51015
#51678

- IDE support:
#58137

- Device network state detection:
#58257

Next steps:
- Port forwarding for `flutter run`
- Supporting `flutter attach`
- `flutter install`, `screenshot` and friends

The Flutter team knows this is needed.  You can show your love for this issue by adding a 👍  instead of adding additional comments.  😄

**Comment by sjapps** [comment]
Sorry for adding a comment but someone said that debugging over wifi is slower (30 seconds vs 5 minutes.). I'd be curious to know where that comes from. Long time iOS dev here and debugging over wifi has become a norm for me. It is slower than cable but I've never seen it that slow.

**Comment by Gebes** [comment]
@sjapps if you are on the go or have an slow internet speed in general, then it can take several minutes.
If building wireless takes quite a long time on iOS, then it also may take the same time on Android (after my own experience)

**Comment by jmagman** [comment]
We should be able to hook up `flutter screenshot` for wireless devices when https://github.com/flutter/flutter/pull/59512 lands.

**Comment by ride4sun** [comment]
Just a idea. You could use scrcpy it works over wifi. Cool tool. I use it
on Mac and Windows and Linux

**Comment by fotiDim** [comment]
To add another use case, I am making an app for an iPhone based robot and connecting with cable is not possible since the lightning port is used to connect to the robot. Wireless debugging was a lifesaver for me.

**Comment by scullabyte** [comment]
Is this planned for the next release?

**Comment by sooxt98** [comment]
welp is there any updates on how to use this on mac & ios via wifi?

**Comment by ThinkDigitalSoftware** [comment]
In my case, I have an unstable connection via USB and it won't show up. I've tried a few cables and don't want to wait or spend on another one when this is the only time I've needed it this year

**Comment by leenin** [comment]
waiting

**Comment by gabrielcosi** [comment]
Is this scheduled for a future update?

**Comment by babyfacehugg** [comment]
Ok thanks

**Comment by acanturgut** [comment]
What is the status of this future?

**Comment by gibahjoe** [comment]
Any timelines on this?

**Comment by sooxt98** [comment]
my guess is gonna wait till this issue close 
https://github.com/flutter/flutter/issues/67580

**Comment by jmagman** [comment]
This is still a priority, but we've diverted resources to get Apple Silicon supported ASAP.  The M1s are shipping now, and iOS devices without ports haven't yet been announced (though I know that's not the only use-case, but it would be the most urgent one for the average user flow).

**Comment by AlexoDz** [comment]
when? 🥺

**Comment by osmanrami** [comment]
I want to add an in app purchase future for my app, it is only supported for real ios devices, and like every other developer, I hate cables ! please we need it asap !

**Comment by edgarfroes** [comment]
My body is ready for this update.

**Comment by jdouitsis** [comment]
Thanks to the flutter team for working on this! Looking forward to when its up and running:)

**Comment by daniel-schulz** [comment]
I have another use case for you to prioritise this issue.  I don't want to charge my devices every day to 100% due to care the batteries life.

**Comment by edgarfroes** [comment]
Sad it's a P4, even though it's one of the most voted issues.

**Comment by jmagman** [comment]
P4 is the default for all bugs and anything we're likely to work on.

**Comment by john7ric** [comment]
Sad that this feature is still  not here.

**Comment by cyc2303** [comment]
Any updates?

**Comment by deandreamatias** [comment]
Please stop commenting on problems things like at:

> any update on this

The best option is to add this emoji :+1: in the initial comment.

**Comment by furkanbahadirozdemir** [comment]
any updates?

**Comment by RA1NO3O** [comment]
connect without cable can avoid damage to the device's battery, come on!

**Comment by thomassth** [comment]
> Please stop commenting on problems things like at:
> 
> > any update on this
> 
> The best option is to add this emoji :+1: in the initial comment.
> 

A little bit more detail on the progress would be helpful. Is it delayed? Still waiting after certain features are completed?

**Comment by jmagman** [comment]
Previous work to get to this point already documented in https://github.com/flutter/flutter/issues/15072#issuecomment-636153777.  The last remaining hurdle on the Flutter side is figuring out a reliable way to get the iOS device IP address to connect to the observatory.  https://github.com/flutter/flutter/pull/58185 came close, though it never worked on my machine.

In the case of a wireless connection, `_forwardPort` needs to swap the `iproxy` port forwarding to instead construct the observatory uri with the device's IP as a host.
https://github.com/flutter/flutter/blob/88f38110553e5c407649ff59f43ac476c5a64aeb/packages/flutter_tools/lib/src/protocol_discovery.dart#L141-L157
```dart
hostUri = hostUri.replace(host: deviceIPAddress);
```

Here's what I have so far, but it's icky:
1. Ship `ideviceinfo` in `bin/cache/artifacts/libimobiledevice`
2. Get the device MAC address.  I can't figure out any way to get the IP address directly from any known domains.
```
$ ideviceinfo -u <UDID> --network -k WiFiAddress
```
3. ping broadcast to populate the arp table
```
$ ping -c 1 192.168.0.255
```
4. Find the arp table entry with the MAC address and parse it out with probably an unreadable regex:
```
$ arp -a -n
...
? (192.168.0.111) at <MAC address is here> on en1 ifscope [ethernet]
...
```
Alternatively we could change it on the device side and have the `Observatory URL on device:` log line to include the device's IP address when `--observatory-host` is `0.0.0.0` or `::`.  That would probably be cleaner.

Since `iproxy --network` happily forwards the wirelessly tethered device port to localhost, I don't really understand why I can't connect to http://127.0.0.1:forwarded-port the same way it's done when it's USB tethered.  I tried this in https://github.com/flutter/flutter/pull/60408 and it didn't work.  I'm not a Mac networking expert.

In the course of working on this, I can tell you the wireless tethering story just within Xcode, without even Flutter, is a mess.  The device disconnects every few minutes, even when it's targeted in an open Xcode project, presumably to save battery life.  Sometimes it looks connected but when you run from Xcode it fails with a `com.apple.dt.MobileDeviceErrorDomain` unknown error.  It forgot its networked connected status maybe every 3rd time I detached from USB, then I had to reattach and re-check the "Connect via network" box.  And then when I was connected with `flutter` (by hardcoding the IP) it would disconnect while I was actually attached with the debugger!  I would have assumed the debug server would have taken whatever power assertions were needed to prevent that, but it didn't.  And I don't see any way for `flutter` to grab such an assertion through lockdown.  The feature is fundamentally flaky.  And I can already picture the flood of "Flutter is broken!" issues that would come from enabling this due to no fault of Flutter tooling.

So that's the current update.  If you have any suggested solutions for the above issues I'd love to hear them.  However, please PLEASE do not reply to this issue with "any updates?" or "me too" or other reasons why this feature should be implemented.  You're adding noise to an issue with 50 participants and untold subscribers.  And you're pushing out actually informative comments (like this one).  Just hit it with a 👍.  I'd like to implement this feature too!

**Comment by iulian0512** [comment]
+1

**Comment by caseyryan** [comment]
Any news on this? Iphone can be easily connected via wifi in xcode, but it would be super useful to do it from Flutter

**Comment by TabooSun** [comment]
I have forked @jmagman [PR](https://github.com/flutter/flutter/pull/60408) [here](https://github.com/TabooSun/flutter/tree/local-stable). It really doesn't work for me as commented [here](https://github.com/flutter/flutter/pull/60408#issuecomment-686215654), however, there is no detail on what isn't working. I tried to dig into it and found out that this is due to the fact that `idevicepair -u <device UUID> pair` is needed to be run and you need to trust this mac on your ios device. Once this is done, you can happily debug wirelessly. I don't see these steps documented in @jmagman [PR](https://github.com/flutter/flutter/pull/60408), so I guess this is the reason it does not work?

However, you need to do this quite frequently, at least from my observation, when any of the devices iPhone or MacBook sleep/restart. I am quite busy recently so my exploration stops here. I guess there should be a way to fix this by looking into the reason why mac can't remember the device.

**Comment by jmagman** [comment]
>  I tried to dig into it and found out that this is due to the fact that `idevicepair -u <device UUID> pair` is needed to be run (while the ios device is connected with cable) and you need to trust this mac on your ios device. 

Thanks for trying it.  Flutter currently assumes the device is already paired in the normal tethered case.  It doesn't try to pair if it's not.  Once I had the device wirelessly set up in Xcode I didn't need to re-trust anything.  It just constantly disconnected and forgot its network connected status, I kept having to re-tether and re-check that box (all in Xcode, this had nothing to do with Flutter).  But I never needed to re-pair once it was paired.

**Comment by epaterlini** [comment]
I am testing locally the "iproxy" program. I developed a very simple TCP server app that runs on my iPad and listens to every interface (0.0.0.0).
I then tried to connect in those ways:
- directly through iPad WiFi's IP address: worked (of course)
- using iproxy with iPad connected through USB (and WiFi disabled): worked
- using iproxy with iPad connected over WiFi: not working (connection refused)

I suspect then that is "iproxy" program that's having trouble forwarding ports over Wi-Fi?
If yes, we can focus on fixing "iproxy" tool first and then retry testing the iOS WiFi debugging.

Let me know what you think guys,
cheers for now.

**Comment by jpangburn** [comment]
@epaterlini looking at the iproxy man page may help.  I know nothing about it, but the Archlinux one is (here)[https://man.archlinux.org/man/extra/libusbmuxd/iproxy.1.en].  Relevant bits may be:

> **-n, --network**
> Connect to network device. The device needs to have WiFi sync enabled and needs to be paired with this computer for this to work. Please note that a device might close connections at any time to save power. This option will make iproxy try to connect to network attched devices only, unless -l is passed too (see below).
> **-l, --local**
> Connect to USB device. This is the default if no option is passed. If used together with -n, iproxy will first attempt a connection to a device attached via USB, and if not available attempt to reach a device via network.
> **-s, --source ADDR**
> Allows setting a source address for the listening socket. The default is 127.0.0.1 (local connections only). While not needed in normal setups, this option allows other than local connections, for example 0.0.0.0 would make the TCP proxy port(s) accessible from other computers in the network.
> WARNING: Use with caution since this could expose a device over the network!

So looks like the default is a USB connection with `-l` and listen on 127.0.0.1.  So may need `-n -s 0.0.0.0` passed as options?  Just wildly throwing a guess out here as I've not played with it at all.

**Comment by epaterlini** [comment]
> 

Hello jpangburn,
yes of course I passed the "-n" flag when connecting through WiFi (otherwise I would get a "No device found" error) :-)
Regarding the "-s" (source) parameter, for me "localhost" (and for the Flutter use case I think) is fine, as I'm trying to forward device's port to localhost and then connect to localhost.
Thank you for your interest, any way :-) do you have some other ideas?

**Comment by jpangburn** [comment]
@epaterlini Sorry I misunderstood.  I thought this line:

> using iproxy with iPad connected over WiFi: not working (connection refused)

meant you were implicitly using `-l` rather than explicitly passing `-n`.  But I see I misunderstood.  Hope you have some success with this!

**Comment by TabooSun** [comment]
> The last remaining hurdle on the Flutter side is figuring out a reliable way to get the iOS device IP address to connect to the observatory.

@jmagman I have a suggestion for this. Since we have already added BonjourService in the Info.plist, i.e. `_dartobservatory._tcp`, we could probably make use of it to get the iOS IP address.

I tried writing a simple Swift CLI app to interact with the [Network framework](https://developer.apple.com/documentation/network) and turns out being able to get the IP address.
```swift
class BonjourService : NSObject, NetServiceDelegate {
	private let domain = "local"
	private let type = "_dartobservatory._tcp"
	private var bonjourBrowser: NWBrowser?
	private var netService: NetService?

	init(name: String) {
		super.init()
		bonjourBrowser = NWBrowser(for: .bonjourWithTXTRecord(type: type, domain: domain), using: NWParameters())
		let netService = NetService(domain: domain, type: type, name: name)
		self.netService = netService
		startDiscovering()
	}

	private func startDiscovering() {
		bonjourBrowser!.browseResultsChangedHandler = {
			(newResults: Set<NWBrowser.Result>, changes: Set<NWBrowser.Result.Change>) in
			self.resolveNetworkInformation()
		}
//		bonjourBrowser.start(queue: DispatchQueue.global())
		bonjourBrowser!.start(queue: DispatchQueue.main)
	}

	private func resolveNetworkInformation() {
		bonjourBrowser!.browseResults.forEach { (result) in
			/*	result.resolve(with: NWBrowser.Resolution.parameters(NWParameters())) { (newResult) in
				print("newResult: \(newResult)")
			}*/
			switch result.endpoint {
				case .hostPort(host: _, port: _):
					break;
				case .service(name: let name, type: let type, domain: let domain, interface: _):
					if netService!.delegate == nil {
						netService!.delegate = self
					}
					netService!.resolve(withTimeout: 10)
					break;
				case .unix(path: _):
					break;
				case .url(_):
					break;
				@unknown default:
					break;
			}
			/*NetService(domain: "local", type: "_dartobservatory._tcp", name: result.endpoint. 
			 )*/
		}
	}

	func netServiceDidResolveAddress(_ sender: NetService) {
		if let addresses = sender.addresses {
			addresses.filter({ data in data.count == 16 }).map { data -> String in
						var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))
						return data.withUnsafeBytes { (ptr) -> String in
							guard let sockaddr_ptr = ptr.baseAddress?.assumingMemoryBound(to: sockaddr.self) else {
								return ""
							}
							let sockaddr = sockaddr_ptr.pointee
							guard getnameinfo(sockaddr_ptr, socklen_t(sockaddr.sa_len), &hostname, socklen_t(hostname.count), nil, 0, NI_NUMERICHOST) == 0 else {
								return ""
							}
							return String(cString: hostname)
						}
					}
					.forEach { s in
						print("\(s)")
					}
		}
	}
}
```

In `main.swift`:
```swift
let bonjourService = BonjourService(name: "com.example.websocketplayground")

while true {
	RunLoop.current.run()
}
```

Now the only missing part that I can't figure out is the way to distinguish between multiple devices while debugging on these devices at the same time. Maybe someone can take it from here?

**Comment by seifkh97** [comment]
any updates on this?

**Comment by saviourdog** [comment]
need it

**Comment by edgarfroes** [comment]
There's a rumor that Apple's gonna have portless iPhones in the next 2 years. The hypothesis is that why should Apple comply with the mandatory replacement of Lightning with USB-C ports if they can just get rid of the port altogether?

**Comment by ahyangnb** [comment]
any news?.

**Comment by jmagman** [comment]
https://github.com/flutter/flutter/pull/118826 was temporarily reverted.

**Comment by jmagman** [comment]
https://github.com/flutter/flutter/pull/118104 is available on the [master channel](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#how-to-change-channels), and we would love some feedback!

### Setup

1. Follow the instructions under "Steps to setup" in the https://github.com/flutter/flutter/pull/118104 description.  There's currently no way to set up wireless debugging from `flutter`, you need to do it from Xcode. 
https://github.com/flutter/flutter/issues/118912
3. Try `flutter devices --device-timeout 5` / `flutter run --device-timeout 5` etc. from the command line.  You may need to adjust that timeout number up depending on your network.  

### Caveats
1. You'll need to add `--device-timeout`  to your `flutter` commands, as networked devices take longer to discover.  We're working on making this a better experience. 
https://github.com/flutter/flutter/issues/118109
https://github.com/flutter/flutter/issues/118113
1. Because of the device timeout issue, when you **Run** from your IDE, it may not discover your wirelessly connected device before timing out.  Try the workaround listed in https://github.com/flutter/flutter/pull/118104 to pass in `--device-timeout` as `args`.
VSCode launch.json example: 
```
{
  ...
  "configurations": [
      {
          "name": "my_app",
          "request": "launch",
          "type": "dart",
          "args": ["--device-timeout", "5"]
      },
      ...
  ]
}
```
3. There is no documentation for iOS wireless debugging yet.  We will add it to the website when the feature is further along.
4. Installing and hot reloading wirelessly over the network is much slower than over a USB tether.  There's not much Flutter can do about that, it's the performance characteristics of Xcode/iOS.
5.  To run your app from Xcode and `flutter attach`, you must add `--observatory-host=0.0.0.0` or `--observatory-host=::0` as a launch argument in your Xcode scheme.  This would also be true for running an add-to-app host app with an embedded Flutter module.  See https://github.com/flutter/flutter/pull/118104 for instructions.
1. If Xcode can't find your device, then Flutter tooling won't be able to either.  If Flutter isn't discovering your device, open `Xcode > Window > Devices and Simulators > Devices` and make sure you see the device is under connected with globe network icon, and Connect via Network is checked.  If either isn't true then you'll have to play with the settings in Xcode until it shows up.
![211061613-9f6f145d-be97-4722-9d2c-f5deee316be5](https://user-images.githubusercontent.com/682784/213819438-1813bbab-1a0f-442c-8b2d-b8d24d68d3cd.png)

Please file new issues, including usability suggestions, as you see them.  Thank you!

**Comment by AmitSonkhiya** [comment]
> #118104 is available on the [master channel](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#how-to-change-channels), and we would love some feedback!

Submitted feedback
https://github.com/flutter/flutter/issues/119493

**Comment by jmagman** [comment]
#118104 is available on [3.7.0-24.0.pre](https://docs.flutter.dev/development/tools/sdk/releases?tab=macos) and higher on the [beta channel](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#how-to-change-channels).  If you are trying to test this you must either be on the master or beta channels, it is not yet available in stable.
